File: Arrays/rotate_array.py
def rotate_array(arr, d):
    # given an unsorted array arr[] of size N, rotate it by D elements (counter clockwise).
    """
    store first element in variable
    shift each element in the array by one position to the left
    place the first element at the end
    do this D times
    """
    if d < 0:
        return "Enter a valid input"
    else:
        for j in range(0,d):
            first_element = arr[0]
            for i in range(1,len(arr)):
                arr[i-1] = arr[i]
            arr[len(arr)-1] = first_element
    
    return arr

def test_cases():
    #test 1: working case - should be [3,5,6,1,2,4]
    result = rotate_array([1,2,4,3,5,6], 3)
    print(result)
    #test 2: 2 elements - result should be [3,5]
    print(rotate_array([5,3],1))
    #test 3:  d = number of elements in array - should be initial array
    print(rotate_array([1,3,4,7,2,5], 6))
    #test 4: d = 0
    print(rotate_array([1,3,4,7,2,5], 0))
    #test 5: d < 0
    print(rotate_array([1,3,4,7,2,5], -1))

test_cases()

File: Arrays/RotateArray.java
class RotateArray {

    /*
     * store first element in variable
    shift each element in the array by one position to the left
    place the first element at the end
    do this D times
    assume array has values because it is unsorted
     */

    public static int[] rotateArray(int[] arr, int d){
        int[] resultArr = arr;
        if(d < 0){
            return null;
        } else {
            for(int i = 0; i < d; i++){
                int firstElem = resultArr[0];
                for(int j = 1; j < resultArr.length; j++){
                    resultArr[j-1] = resultArr[j];
                }
                resultArr[resultArr.length-1] = firstElem;
            }
        }
        return resultArr;
    }

    static String toString(int[] arr){
        String result = "";
        for(int i : arr){
            result = result + i + " ";
        }
        return result;
    }
    public static void main(String[] args){
        //test 1 - normal working case:
        int[] result = rotateArray(new int[]{1,2,4,3,5,6}, 3);
        System.out.println(toString(result)); // 3 5 6 1 2 4
        // test 2 - d less than 0:
        System.out.println(rotateArray(new int[]{1,2,3,4,5,6}, -1)); // null
        //test 3 - d equals # of elements:
        System.out.println(toString(rotateArray(new int[]{1,2,4,3,5,6}, 6))); // 1,2,4,3,5,6
        //test 4 - d > # of elements:
        System.out.println(toString(rotateArray(new int[]{1,2,4,3,5,6}, 7))); // 2,4,3,5,6,1
        //test 5 - 2 elements:
        System.out.println(toString(rotateArray(new int[]{5,3}, 1))); // 3,5
    }
}